### PR TITLE
Add observability (metrics and tracing) to RabbitMQ connector

### PR DESCRIPTION
**Pull request by kaneeldias**
## Purpose
Add observability (metrics and tracing) to RabbitMQ connector

### CODE REVIEW

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/MessageDispatcher.java [hunk 1]
```diff
@@ -221,17 +232,69 @@ private void handleError(byte[] message, long deliveryTag, AMQP.BasicProperties
...
+    }
+
+    private void executeResource(CountDownLatch countDownLatch, byte[] message, ObjectValue messageObjectValue,
+                                 Object forContent) {
+        if (ObserveUtils.isTracingEnabled()) {
+            Map<String, Object> observerProperties = getNewObserverContextInProperties();
+            runtime.invokeMethodAsync(service, RabbitMQConstants.FUNC_ON_MESSAGE,
+                                      new RabbitMQResourceCallback(countDownLatch, channel, queueName, message.length),
+                                      observerProperties, messageObjectValue, true, forContent, true);
+            return;
```
Comment (aashikam) [inline stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/MessageDispatcher.java]: Can make these two methods more simpler. It does the same thing and runtime.invokeMethodAsync() is called 4 times inside these two methods :) I think we can make it one. Except for the countdownLatch and the RabbitMQResourceCallback, everything else goes into the Object.. args parameter of the invokeMethodAsync() method. This should give you a start to refactoring.

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/observability/RabbitMQObservabilityConstants.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    static final String TAG_ACK_TYPE = "ack_type";
+    static final String TAG_CONSUME_TYPE = "consume_type";
+
+    public static final String ERROR_TYPE_CONNECTION = "connection";
+    public static final String ERROR_TYPE_CONNECTION_CLOSE = "connection_close";
+    public static final String ERROR_TYPE_CONNECTION_ABORT = "connection_abort";
+    public static final String ERROR_TYPE_PUBLISH = "publish";
+    public static final String ERROR_TYPE_CONSUME = "consume";
+    public static final String ERROR_TYPE_CHANNEL_CREATE = "channel_create";
+    public static final String ERROR_TYPE_CHANNEL_CLOSE = "create_channel";
```
Comment (chamil321) [inline stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/observability/RabbitMQObservabilityConstants.java]: I believe you meant to say `channel_close` here :) 
```suggestion
    public static final String ERROR_TYPE_CHANNEL_CLOSE = "channel_close";
```

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/RabbitMQErrorResourceCallback.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import org.ballerinalang.jvm.services.ErrorHandlerUtils;
+import org.ballerinalang.jvm.values.ErrorValue;
+import org.ballerinalang.jvm.values.connector.CallableUnitCallback;
+import org.ballerinalang.messaging.rabbitmq.observability.RabbitMQMetricsUtil;
+import org.ballerinalang.messaging.rabbitmq.observability.RabbitMQObservabilityConstants;
+
+import java.util.concurrent.CountDownLatch;
+
+/**
+ * The error resource call back implementation for RabbitMQ async consumer.
```
Comment (chamil321) [inline stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/RabbitMQErrorResourceCallback.java]:
```suggestion
 * The error resource callback implementation for RabbitMQ async consumer.
```

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/MessageDispatcher.java [hunk 4]
```diff
@@ -221,17 +232,69 @@ private void handleError(byte[] message, long deliveryTag, AMQP.BasicProperties
...
+                                  new RabbitMQResourceCallback(countDownLatch, channel, queueName, message.length),
+                                  messageObjectValue, true, forContent, true);
+    }
+
+    private void executeOnErrorResource(CountDownLatch countDownLatch, ObjectValue messageObjectValue, ErrorValue error,
+                                        AMQP.BasicProperties properties) {
+        if (ObserveUtils.isTracingEnabled()) {
+            Map<String, Object> observerProperties = getNewObserverContextInProperties();
+            runtime.invokeMethodAsync(service, RabbitMQConstants.FUNC_ON_ERROR,
+                                      new RabbitMQErrorResourceCallback(countDownLatch), observerProperties,
```
Comment (chamil321) [inline stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/MessageDispatcher.java]:
```suggestion
                                      new RabbitMQErrorResourceCallback(countDownLatch), getNewObserverContextInProperties(),
```
As the variable has no any other usages.

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/observability/RabbitMQObservabilityUtil.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * @since 1.2.0
+ */
+public class RabbitMQObservabilityUtil {
+
+    public static String getServerUrl(Connection connection) {
+        return connection.toString();
+    }
+
+    public static String getChannel(Channel channel) {
```
Comment (chamil321) [inline stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/observability/RabbitMQObservabilityUtil.java]: Do we have any usage of this method? If not, let's remove it

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/observability/RabbitMQObservabilityUtil.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    public static String getServerUrl(Connection connection) {
+        return connection.toString();
+    }
+
+    public static String getQueueName(ObjectValue service) {
+        return (String) service.getNativeData(RabbitMQConstants.QUEUE_NAME);
+    }
+}
+
```
Comment (kalaiyarasiganeshalingam) [inline stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/observability/RabbitMQObservabilityUtil.java]: Remove this line

**Pull request merged**